Controller manager crashing with std::runtime_error

**Describe the bug**
In current testing the controller manager crashes with a

```
[ros2_control_node-1] terminate called after throwing an instance of 'std::runtime_error'
[ros2_control_node-1]   what():  can't compare times with different time sources
```

**To Reproduce**
Steps to reproduce the behavior:
1. build `ros2_control`, `ros2_controllers` and `ros2_control_demos` from source with latest master. I haven't checked with older versions at the moment.
2. launch a demo example e.g. `ros2 launch ros2_control_demo_example_1 rrbot.launch.py`

**Expected behavior**
It should work :-)

**Stacktrace**
```
[ros2_control_node-1] terminate called after throwing an instance of 'std::runtime_error'
[ros2_control_node-1]   what():  can't compare times with different time sources
[ros2_control_node-1] Stack trace (most recent call last) in thread 101078:
[ros2_control_node-1] #13   Object "", at 0xffffffffffffffff, in
[ros2_control_node-1] #12   Source "../sysdeps/unix/sysv/linux/x86_64/clone3.S", line 81, in __clone3 [0x7f57959f3a3f]
[ros2_control_node-1] #11   Source "./nptl/pthread_create.c", line 442, in start_thread [0x7f5795961ac2]
[ros2_control_node-1] #10   Object "/usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.30", at 0x7f5795bf1252, in
[ros2_control_node-1] #9    Source "/home/ubuntu/checkout/ros2_control/colcon_ws/src/ros2_control/controller_manager/src/ros2_control_node.cpp", line 81, in operator() [0x55ad79027088]
[ros2_control_node-1]          79:         // execute update loop
[ros2_control_node-1]          80:         cm->read(cm->now(), measured_period);
[ros2_control_node-1]       >  81:         cm->update(cm->now(), measured_period);
[ros2_control_node-1]          82:         cm->write(cm->now(), measured_period);
[ros2_control_node-1]          83:
[ros2_control_node-1]          84:         // wait until we hit the end of the period
[INFO] [spawner-3]: process has finished cleanly [pid 101052]
[INFO] [rviz2-4]: process started with pid [101104]
[rviz2-4] QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-ubuntu'
[ros2_control_node-1] #8    Source "/home/ubuntu/checkout/ros2_control/colcon_ws/src/ros2_control/controller_manager/src/controller_manager.cpp", line 2046, in update [0x7f57960977ac]
[ros2_control_node-1]        2043:         run_controller_at_cm_rate ? period
[ros2_control_node-1]        2044:                                   : rclcpp::Duration::from_seconds((1.0 / controller_update_rate));
[ros2_control_node-1]        2045:
[ros2_control_node-1]       >2046:       bool controller_go = (time == rclcpp::Time(0)) ||
[ros2_control_node-1]        2047:                            (time.seconds() >= loaded_controller.next_update_cycle_time->seconds());
[ros2_control_node-1]        2048:
[ros2_control_node-1]        2049:       RCLCPP_DEBUG(
[ros2_control_node-1] #7    Object "/opt/ros/rolling/lib/librclcpp.so", at 0x7f5795e45b09, in
[ros2_control_node-1] #6    Object "/usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.30", at 0x7f5795bc34d7, in __cxa_throw
[ros2_control_node-1] #5    Object "/usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.30", at 0x7f5795bc3276, in std::terminate()
[ros2_control_node-1] #4    Object "/usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.30", at 0x7f5795bc320b, in
[ros2_control_node-1] #3    Object "/usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.30", at 0x7f5795bb7b9d, in
[ros2_control_node-1] #2    Source "./stdlib/abort.c", line 79, in abort [0x7f57958f57f2]
[ros2_control_node-1] #1    Source "../sysdeps/posix/raise.c", line 26, in raise [0x7f579590f475]
[ros2_control_node-1] #0  | Source "./nptl/pthread_kill.c", line 89, in __pthread_kill_internal
[ros2_control_node-1]     | Source "./nptl/pthread_kill.c", line 78, in __pthread_kill_implementation
[ros2_control_node-1]       Source "./nptl/pthread_kill.c", line 44, in __pthread_kill [0x7f57959639fc]
[ros2_control_node-1] Aborted (Signal sent by tkill() 101048 1000)
[ERROR] [ros2_control_node-1]: process has died
```

**Environment (please complete the following information):**
 - OS: Ubuntu 22.04
 - Version ROS Rolling with ros2_control built from source

--- a/controller_manager/src/controller_manager.cpp
+++ b/controller_manager/src/controller_manager.cpp
@@ -565,7 +565,8 @@ controller_interface::ControllerInterfaceBaseSharedPtr ControllerManager::load_c
   controller_spec.c = controller;
   controller_spec.info.name = controller_name;
   controller_spec.info.type = controller_type;
-  controller_spec.next_update_cycle_time = std::make_shared<rclcpp::Time>(0);
+  controller_spec.next_update_cycle_time = std::make_shared<rclcpp::Time>(
+    0, 0, this->get_node_clock_interface()->get_clock()->get_clock_type());
 
   return add_controller_impl(controller_spec);
 }
@@ -1457,7 +1458,8 @@ void ControllerManager::activate_controllers(
     auto controller = found_it->c;
     auto controller_name = found_it->info.name;
     // reset the next update cycle time for newly activated controllers
-    *found_it->next_update_cycle_time = rclcpp::Time(0);
+    *found_it->next_update_cycle_time =
+      rclcpp::Time(0, 0, this->get_node_clock_interface()->get_clock()->get_clock_type());
 
     bool assignment_successful = true;
     // assign command interfaces to the controller
@@ -2043,8 +2045,10 @@ controller_interface::return_type ControllerManager::update(
         run_controller_at_cm_rate ? period
                                   : rclcpp::Duration::from_seconds((1.0 / controller_update_rate));
 
-      bool controller_go = (time == rclcpp::Time(0)) ||
-                           (time.seconds() >= loaded_controller.next_update_cycle_time->seconds());
+      bool controller_go =
+        (time ==
+         rclcpp::Time(0, 0, this->get_node_clock_interface()->get_clock()->get_clock_type())) ||
+        (time.seconds() >= loaded_controller.next_update_cycle_time->seconds());
 
       RCLCPP_DEBUG(
         get_logger(), "update_loop_counter: '%d ' controller_go: '%s ' controller_name: '%s '",
@@ -2055,7 +2059,9 @@ controller_interface::return_type ControllerManager::update(
       {
         auto controller_ret = loaded_controller.c->update(time, controller_period);
 
-        if (*loaded_controller.next_update_cycle_time == rclcpp::Time(0))
+        if (
+          *loaded_controller.next_update_cycle_time ==
+          rclcpp::Time(0, 0, this->get_node_clock_interface()->get_clock()->get_clock_type()))
         {
           *loaded_controller.next_update_cycle_time = time;
         }

--- a/controller_manager/test/controller_manager_test_common.hpp
+++ b/controller_manager/test/controller_manager_test_common.hpp
@@ -41,7 +41,6 @@
 
 namespace
 {
-const auto TIME = rclcpp::Time(0);
 const auto PERIOD = rclcpp::Duration::from_seconds(0.01);
 const auto STRICT = controller_manager_msgs::srv::SwitchController::Request::STRICT;
 const auto BEST_EFFORT = controller_manager_msgs::srv::SwitchController::Request::BEST_EFFORT;
@@ -119,7 +118,9 @@ public:
       {
         while (run_updater_)
         {
-          cm_->update(rclcpp::Time(0), rclcpp::Duration::from_seconds(0.01));
+          cm_->update(
+            rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+            rclcpp::Duration::from_seconds(0.01));
           std::this_thread::sleep_for(std::chrono::milliseconds(10));
         }
       });
@@ -177,9 +178,15 @@ public:
       std::chrono::milliseconds(10),
       [&]()
       {
-        cm_->read(TIME, PERIOD);
-        cm_->update(TIME, PERIOD);
-        cm_->write(TIME, PERIOD);
+        cm_->read(
+          rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+          PERIOD);
+        cm_->update(
+          rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+          PERIOD);
+        cm_->write(
+          rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+          PERIOD);
       });
 
     executor_->add_node(cm_);
@@ -206,7 +213,9 @@ public:
       while (service_executor.spin_until_future_complete(result, std::chrono::milliseconds(50)) !=
              rclcpp::FutureReturnCode::SUCCESS)
       {
-        cm_->update(rclcpp::Time(0), rclcpp::Duration::from_seconds(0.01));
+        cm_->update(
+          rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+          rclcpp::Duration::from_seconds(0.01));
       }
     }
     else

--- a/controller_manager/test/controller_manager_test_common.hpp
+++ b/controller_manager/test/controller_manager_test_common.hpp
@@ -100,6 +100,7 @@ public:
         cm_->robot_description_callback(msg);
       }
     }
+    time_ = rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type());
   }
 
   static void SetUpTestCase() { rclcpp::init(0, nullptr); }
@@ -118,9 +119,7 @@ public:
       {
         while (run_updater_)
         {
-          cm_->update(
-            rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-            rclcpp::Duration::from_seconds(0.01));
+          cm_->update(time_, rclcpp::Duration::from_seconds(0.01));
           std::this_thread::sleep_for(std::chrono::milliseconds(10));
         }
       });
@@ -158,6 +157,7 @@ public:
   bool run_updater_;
   const std::string robot_description_;
   const bool pass_urdf_as_parameter_;
+  rclcpp::Time time_;
 };
 
 class TestControllerManagerSrvs
@@ -178,15 +178,9 @@ public:
       std::chrono::milliseconds(10),
       [&]()
       {
-        cm_->read(
-          rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-          PERIOD);
-        cm_->update(
-          rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-          PERIOD);
-        cm_->write(
-          rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-          PERIOD);
+        cm_->read(time_, PERIOD);
+        cm_->update(time_, PERIOD);
+        cm_->write(time_, PERIOD);
       });
 
     executor_->add_node(cm_);
@@ -213,9 +207,7 @@ public:
       while (service_executor.spin_until_future_complete(result, std::chrono::milliseconds(50)) !=
              rclcpp::FutureReturnCode::SUCCESS)
       {
-        cm_->update(
-          rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-          rclcpp::Duration::from_seconds(0.01));
+        cm_->update(time_, rclcpp::Duration::from_seconds(0.01));
       }
     }
     else

--- a/controller_manager/test/test_controller_manager.cpp
+++ b/controller_manager/test/test_controller_manager.cpp
@@ -96,9 +96,7 @@ TEST_P(TestControllerManagerWithStrictness, controller_lifecycle)
 
   EXPECT_EQ(
     controller_interface::return_type::OK,
-    cm_->update(
-      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-      rclcpp::Duration::from_seconds(0.01)));
+    cm_->update(time_, rclcpp::Duration::from_seconds(0.01)));
   EXPECT_EQ(0u, test_controller->internal_counter)
     << "Update should not reach an unconfigured controller";
 
@@ -110,9 +108,7 @@ TEST_P(TestControllerManagerWithStrictness, controller_lifecycle)
   cm_->configure_controller(TEST_CONTROLLER2_NAME);
   EXPECT_EQ(
     controller_interface::return_type::OK,
-    cm_->update(
-      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-      rclcpp::Duration::from_seconds(0.01)));
+    cm_->update(time_, rclcpp::Duration::from_seconds(0.01)));
   EXPECT_EQ(0u, test_controller->internal_counter) << "Controller is not started";
   EXPECT_EQ(0u, test_controller2->internal_counter) << "Controller is not started";
 
@@ -127,9 +123,7 @@ TEST_P(TestControllerManagerWithStrictness, controller_lifecycle)
 
   EXPECT_EQ(
     controller_interface::return_type::OK,
-    cm_->update(
-      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-      rclcpp::Duration::from_seconds(0.01)));
+    cm_->update(time_, rclcpp::Duration::from_seconds(0.01)));
   EXPECT_EQ(0u, test_controller2->internal_counter) << "Controller is started at the end of update";
   {
     ControllerManagerRunner cm_runner(this);
@@ -138,9 +132,7 @@ TEST_P(TestControllerManagerWithStrictness, controller_lifecycle)
 
   EXPECT_EQ(
     controller_interface::return_type::OK,
-    cm_->update(
-      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-      rclcpp::Duration::from_seconds(0.01)));
+    cm_->update(time_, rclcpp::Duration::from_seconds(0.01)));
   EXPECT_GE(test_controller2->internal_counter, test_param.expected_counter);
 
   // Start the real test controller, will take effect at the end of the update function
@@ -155,9 +147,7 @@ TEST_P(TestControllerManagerWithStrictness, controller_lifecycle)
 
   EXPECT_EQ(
     controller_interface::return_type::OK,
-    cm_->update(
-      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-      rclcpp::Duration::from_seconds(0.01)));
+    cm_->update(time_, rclcpp::Duration::from_seconds(0.01)));
   EXPECT_EQ(0u, test_controller->internal_counter) << "Controller is started at the end of update";
   {
     ControllerManagerRunner cm_runner(this);
@@ -167,9 +157,7 @@ TEST_P(TestControllerManagerWithStrictness, controller_lifecycle)
 
   EXPECT_EQ(
     controller_interface::return_type::OK,
-    cm_->update(
-      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-      rclcpp::Duration::from_seconds(0.01)));
+    cm_->update(time_, rclcpp::Duration::from_seconds(0.01)));
   EXPECT_GE(test_controller->internal_counter, 1u);
   size_t last_internal_counter = test_controller->internal_counter;
 
@@ -185,9 +173,7 @@ TEST_P(TestControllerManagerWithStrictness, controller_lifecycle)
 
   EXPECT_EQ(
     controller_interface::return_type::OK,
-    cm_->update(
-      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-      rclcpp::Duration::from_seconds(0.01)));
+    cm_->update(time_, rclcpp::Duration::from_seconds(0.01)));
   EXPECT_EQ(last_internal_counter + 1u, test_controller->internal_counter)
     << "Controller is stopped at the end of update, so it should have done one more update";
   {
@@ -222,9 +208,7 @@ TEST_P(TestControllerManagerWithStrictness, per_controller_update_rate)
 
   EXPECT_EQ(
     controller_interface::return_type::OK,
-    cm_->update(
-      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-      rclcpp::Duration::from_seconds(0.01)));
+    cm_->update(time_, rclcpp::Duration::from_seconds(0.01)));
   EXPECT_EQ(0u, test_controller->internal_counter)
     << "Update should not reach an unconfigured controller";
 
@@ -236,9 +220,7 @@ TEST_P(TestControllerManagerWithStrictness, per_controller_update_rate)
   cm_->configure_controller(test_controller::TEST_CONTROLLER_NAME);
   EXPECT_EQ(
     controller_interface::return_type::OK,
-    cm_->update(
-      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-      rclcpp::Duration::from_seconds(0.01)));
+    cm_->update(time_, rclcpp::Duration::from_seconds(0.01)));
   EXPECT_EQ(0u, test_controller->internal_counter) << "Controller is not started";
 
   EXPECT_EQ(lifecycle_msgs::msg::State::PRIMARY_STATE_INACTIVE, test_controller->get_state().id());
@@ -255,9 +237,7 @@ TEST_P(TestControllerManagerWithStrictness, per_controller_update_rate)
 
   EXPECT_EQ(
     controller_interface::return_type::OK,
-    cm_->update(
-      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-      rclcpp::Duration::from_seconds(0.01)));
+    cm_->update(time_, rclcpp::Duration::from_seconds(0.01)));
   EXPECT_EQ(0u, test_controller->internal_counter) << "Controller is started at the end of update";
   {
     ControllerManagerRunner cm_runner(this);
@@ -271,9 +251,7 @@ TEST_P(TestControllerManagerWithStrictness, per_controller_update_rate)
   {
     EXPECT_EQ(
       controller_interface::return_type::OK,
-      cm_->update(
-        rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-        rclcpp::Duration::from_seconds(0.01)));
+      cm_->update(time_, rclcpp::Duration::from_seconds(0.01)));
   }
   EXPECT_GE(test_controller->internal_counter, 1u);
   EXPECT_EQ(test_controller->get_update_rate(), 4u);
@@ -308,9 +286,7 @@ TEST_P(TestControllerManagerWithUpdateRates, per_controller_equal_and_higher_upd
   EXPECT_EQ(2, test_controller.use_count());
   EXPECT_EQ(
     controller_interface::return_type::OK,
-    cm_->update(
-      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-      rclcpp::Duration::from_seconds(0.01)));
+    cm_->update(time_, rclcpp::Duration::from_seconds(0.01)));
   EXPECT_EQ(last_internal_counter, test_controller->internal_counter)
     << "Update should not reach an unconfigured controller";
 
@@ -323,9 +299,7 @@ TEST_P(TestControllerManagerWithUpdateRates, per_controller_equal_and_higher_upd
   cm_->configure_controller(test_controller::TEST_CONTROLLER_NAME);
   EXPECT_EQ(
     controller_interface::return_type::OK,
-    cm_->update(
-      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-      rclcpp::Duration::from_seconds(0.01)));
+    cm_->update(time_, rclcpp::Duration::from_seconds(0.01)));
   EXPECT_EQ(last_internal_counter, test_controller->internal_counter)
     << "Controller is not started";
   EXPECT_EQ(lifecycle_msgs::msg::State::PRIMARY_STATE_INACTIVE, test_controller->get_state().id());
@@ -342,9 +316,7 @@ TEST_P(TestControllerManagerWithUpdateRates, per_controller_equal_and_higher_upd
 
   EXPECT_EQ(
     controller_interface::return_type::OK,
-    cm_->update(
-      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-      rclcpp::Duration::from_seconds(0.01)));
+    cm_->update(time_, rclcpp::Duration::from_seconds(0.01)));
   EXPECT_EQ(last_internal_counter, test_controller->internal_counter)
     << "Controller is started at the end of update";
   {
@@ -360,9 +332,7 @@ TEST_P(TestControllerManagerWithUpdateRates, per_controller_equal_and_higher_upd
   {
     EXPECT_EQ(
       controller_interface::return_type::OK,
-      cm_->update(
-        rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-        rclcpp::Duration::from_seconds(0.01)));
+      cm_->update(time_, rclcpp::Duration::from_seconds(0.01)));
     loop_rate.sleep();
   }
   // if we do 2 times of the controller_manager update rate, the internal counter should be
@@ -378,10 +348,7 @@ TEST_P(TestControllerManagerWithUpdateRates, per_controller_equal_and_higher_upd
     << "switch_controller should be blocking until next update cycle";
 
   EXPECT_EQ(
-    controller_interface::return_type::OK,
-    cm_->update(
-      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-      rclcpp::Duration::from_seconds(0.01)))
+    controller_interface::return_type::OK, cm_->update(time_, rclcpp::Duration::from_seconds(0.01)))
     << "Controller is stopped at the end of update, so it should have done one more update";
   {
     ControllerManagerRunner cm_runner(this);
@@ -425,9 +392,7 @@ TEST_P(TestControllerUpdateRates, check_the_controller_update_rate)
   cm_->configure_controller(test_controller::TEST_CONTROLLER_NAME);
   EXPECT_EQ(
     controller_interface::return_type::OK,
-    cm_->update(
-      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-      rclcpp::Duration::from_seconds(0.01)));
+    cm_->update(time_, rclcpp::Duration::from_seconds(0.01)));
   EXPECT_EQ(0u, test_controller->internal_counter) << "Controller is not started";
 
   EXPECT_EQ(lifecycle_msgs::msg::State::PRIMARY_STATE_INACTIVE, test_controller->get_state().id());
@@ -445,9 +410,7 @@ TEST_P(TestControllerUpdateRates, check_the_controller_update_rate)
 
   EXPECT_EQ(
     controller_interface::return_type::OK,
-    cm_->update(
-      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-      rclcpp::Duration::from_seconds(0.01)));
+    cm_->update(time_, rclcpp::Duration::from_seconds(0.01)));
   EXPECT_EQ(0u, test_controller->internal_counter) << "Controller is started at the end of update";
   {
     ControllerManagerRunner cm_runner(this);
@@ -461,8 +424,7 @@ TEST_P(TestControllerUpdateRates, check_the_controller_update_rate)
   const auto controller_update_rate = test_controller->get_update_rate();
 
   const auto initial_counter = test_controller->internal_counter;
-  rclcpp::Time time =
-    rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type());
+  rclcpp::Time time = time_;
   for (size_t update_counter = 0; update_counter <= 10 * cm_update_rate; ++update_counter)
   {
     EXPECT_EQ(

--- a/controller_manager/test/test_controller_manager.cpp
+++ b/controller_manager/test/test_controller_manager.cpp
@@ -96,7 +96,9 @@ TEST_P(TestControllerManagerWithStrictness, controller_lifecycle)
 
   EXPECT_EQ(
     controller_interface::return_type::OK,
-    cm_->update(rclcpp::Time(0), rclcpp::Duration::from_seconds(0.01)));
+    cm_->update(
+      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+      rclcpp::Duration::from_seconds(0.01)));
   EXPECT_EQ(0u, test_controller->internal_counter)
     << "Update should not reach an unconfigured controller";
 
@@ -108,7 +110,9 @@ TEST_P(TestControllerManagerWithStrictness, controller_lifecycle)
   cm_->configure_controller(TEST_CONTROLLER2_NAME);
   EXPECT_EQ(
     controller_interface::return_type::OK,
-    cm_->update(rclcpp::Time(0), rclcpp::Duration::from_seconds(0.01)));
+    cm_->update(
+      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+      rclcpp::Duration::from_seconds(0.01)));
   EXPECT_EQ(0u, test_controller->internal_counter) << "Controller is not started";
   EXPECT_EQ(0u, test_controller2->internal_counter) << "Controller is not started";
 
@@ -123,7 +127,9 @@ TEST_P(TestControllerManagerWithStrictness, controller_lifecycle)
 
   EXPECT_EQ(
     controller_interface::return_type::OK,
-    cm_->update(rclcpp::Time(0), rclcpp::Duration::from_seconds(0.01)));
+    cm_->update(
+      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+      rclcpp::Duration::from_seconds(0.01)));
   EXPECT_EQ(0u, test_controller2->internal_counter) << "Controller is started at the end of update";
   {
     ControllerManagerRunner cm_runner(this);
@@ -132,7 +138,9 @@ TEST_P(TestControllerManagerWithStrictness, controller_lifecycle)
 
   EXPECT_EQ(
     controller_interface::return_type::OK,
-    cm_->update(rclcpp::Time(0), rclcpp::Duration::from_seconds(0.01)));
+    cm_->update(
+      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+      rclcpp::Duration::from_seconds(0.01)));
   EXPECT_GE(test_controller2->internal_counter, test_param.expected_counter);
 
   // Start the real test controller, will take effect at the end of the update function
@@ -147,7 +155,9 @@ TEST_P(TestControllerManagerWithStrictness, controller_lifecycle)
 
   EXPECT_EQ(
     controller_interface::return_type::OK,
-    cm_->update(rclcpp::Time(0), rclcpp::Duration::from_seconds(0.01)));
+    cm_->update(
+      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+      rclcpp::Duration::from_seconds(0.01)));
   EXPECT_EQ(0u, test_controller->internal_counter) << "Controller is started at the end of update";
   {
     ControllerManagerRunner cm_runner(this);
@@ -157,7 +167,9 @@ TEST_P(TestControllerManagerWithStrictness, controller_lifecycle)
 
   EXPECT_EQ(
     controller_interface::return_type::OK,
-    cm_->update(rclcpp::Time(0), rclcpp::Duration::from_seconds(0.01)));
+    cm_->update(
+      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+      rclcpp::Duration::from_seconds(0.01)));
   EXPECT_GE(test_controller->internal_counter, 1u);
   size_t last_internal_counter = test_controller->internal_counter;
 
@@ -173,7 +185,9 @@ TEST_P(TestControllerManagerWithStrictness, controller_lifecycle)
 
   EXPECT_EQ(
     controller_interface::return_type::OK,
-    cm_->update(rclcpp::Time(0), rclcpp::Duration::from_seconds(0.01)));
+    cm_->update(
+      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+      rclcpp::Duration::from_seconds(0.01)));
   EXPECT_EQ(last_internal_counter + 1u, test_controller->internal_counter)
     << "Controller is stopped at the end of update, so it should have done one more update";
   {
@@ -208,7 +222,9 @@ TEST_P(TestControllerManagerWithStrictness, per_controller_update_rate)
 
   EXPECT_EQ(
     controller_interface::return_type::OK,
-    cm_->update(rclcpp::Time(0), rclcpp::Duration::from_seconds(0.01)));
+    cm_->update(
+      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+      rclcpp::Duration::from_seconds(0.01)));
   EXPECT_EQ(0u, test_controller->internal_counter)
     << "Update should not reach an unconfigured controller";
 
@@ -220,7 +236,9 @@ TEST_P(TestControllerManagerWithStrictness, per_controller_update_rate)
   cm_->configure_controller(test_controller::TEST_CONTROLLER_NAME);
   EXPECT_EQ(
     controller_interface::return_type::OK,
-    cm_->update(rclcpp::Time(0), rclcpp::Duration::from_seconds(0.01)));
+    cm_->update(
+      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+      rclcpp::Duration::from_seconds(0.01)));
   EXPECT_EQ(0u, test_controller->internal_counter) << "Controller is not started";
 
   EXPECT_EQ(lifecycle_msgs::msg::State::PRIMARY_STATE_INACTIVE, test_controller->get_state().id());
@@ -237,7 +255,9 @@ TEST_P(TestControllerManagerWithStrictness, per_controller_update_rate)
 
   EXPECT_EQ(
     controller_interface::return_type::OK,
-    cm_->update(rclcpp::Time(0), rclcpp::Duration::from_seconds(0.01)));
+    cm_->update(
+      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+      rclcpp::Duration::from_seconds(0.01)));
   EXPECT_EQ(0u, test_controller->internal_counter) << "Controller is started at the end of update";
   {
     ControllerManagerRunner cm_runner(this);
@@ -251,7 +271,9 @@ TEST_P(TestControllerManagerWithStrictness, per_controller_update_rate)
   {
     EXPECT_EQ(
       controller_interface::return_type::OK,
-      cm_->update(rclcpp::Time(0), rclcpp::Duration::from_seconds(0.01)));
+      cm_->update(
+        rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+        rclcpp::Duration::from_seconds(0.01)));
   }
   EXPECT_GE(test_controller->internal_counter, 1u);
   EXPECT_EQ(test_controller->get_update_rate(), 4u);
@@ -286,7 +308,9 @@ TEST_P(TestControllerManagerWithUpdateRates, per_controller_equal_and_higher_upd
   EXPECT_EQ(2, test_controller.use_count());
   EXPECT_EQ(
     controller_interface::return_type::OK,
-    cm_->update(rclcpp::Time(0), rclcpp::Duration::from_seconds(0.01)));
+    cm_->update(
+      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+      rclcpp::Duration::from_seconds(0.01)));
   EXPECT_EQ(last_internal_counter, test_controller->internal_counter)
     << "Update should not reach an unconfigured controller";
 
@@ -299,7 +323,9 @@ TEST_P(TestControllerManagerWithUpdateRates, per_controller_equal_and_higher_upd
   cm_->configure_controller(test_controller::TEST_CONTROLLER_NAME);
   EXPECT_EQ(
     controller_interface::return_type::OK,
-    cm_->update(rclcpp::Time(0), rclcpp::Duration::from_seconds(0.01)));
+    cm_->update(
+      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+      rclcpp::Duration::from_seconds(0.01)));
   EXPECT_EQ(last_internal_counter, test_controller->internal_counter)
     << "Controller is not started";
   EXPECT_EQ(lifecycle_msgs::msg::State::PRIMARY_STATE_INACTIVE, test_controller->get_state().id());
@@ -316,7 +342,9 @@ TEST_P(TestControllerManagerWithUpdateRates, per_controller_equal_and_higher_upd
 
   EXPECT_EQ(
     controller_interface::return_type::OK,
-    cm_->update(rclcpp::Time(0), rclcpp::Duration::from_seconds(0.01)));
+    cm_->update(
+      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+      rclcpp::Duration::from_seconds(0.01)));
   EXPECT_EQ(last_internal_counter, test_controller->internal_counter)
     << "Controller is started at the end of update";
   {
@@ -332,7 +360,9 @@ TEST_P(TestControllerManagerWithUpdateRates, per_controller_equal_and_higher_upd
   {
     EXPECT_EQ(
       controller_interface::return_type::OK,
-      cm_->update(rclcpp::Time(0), rclcpp::Duration::from_seconds(0.01)));
+      cm_->update(
+        rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+        rclcpp::Duration::from_seconds(0.01)));
     loop_rate.sleep();
   }
   // if we do 2 times of the controller_manager update rate, the internal counter should be
@@ -349,7 +379,9 @@ TEST_P(TestControllerManagerWithUpdateRates, per_controller_equal_and_higher_upd
 
   EXPECT_EQ(
     controller_interface::return_type::OK,
-    cm_->update(rclcpp::Time(0), rclcpp::Duration::from_seconds(0.01)))
+    cm_->update(
+      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+      rclcpp::Duration::from_seconds(0.01)))
     << "Controller is stopped at the end of update, so it should have done one more update";
   {
     ControllerManagerRunner cm_runner(this);
@@ -393,7 +425,9 @@ TEST_P(TestControllerUpdateRates, check_the_controller_update_rate)
   cm_->configure_controller(test_controller::TEST_CONTROLLER_NAME);
   EXPECT_EQ(
     controller_interface::return_type::OK,
-    cm_->update(rclcpp::Time(0), rclcpp::Duration::from_seconds(0.01)));
+    cm_->update(
+      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+      rclcpp::Duration::from_seconds(0.01)));
   EXPECT_EQ(0u, test_controller->internal_counter) << "Controller is not started";
 
   EXPECT_EQ(lifecycle_msgs::msg::State::PRIMARY_STATE_INACTIVE, test_controller->get_state().id());
@@ -411,7 +445,9 @@ TEST_P(TestControllerUpdateRates, check_the_controller_update_rate)
 
   EXPECT_EQ(
     controller_interface::return_type::OK,
-    cm_->update(rclcpp::Time(0), rclcpp::Duration::from_seconds(0.01)));
+    cm_->update(
+      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+      rclcpp::Duration::from_seconds(0.01)));
   EXPECT_EQ(0u, test_controller->internal_counter) << "Controller is started at the end of update";
   {
     ControllerManagerRunner cm_runner(this);
@@ -425,7 +461,8 @@ TEST_P(TestControllerUpdateRates, check_the_controller_update_rate)
   const auto controller_update_rate = test_controller->get_update_rate();
 
   const auto initial_counter = test_controller->internal_counter;
-  rclcpp::Time time = rclcpp::Time(0);
+  rclcpp::Time time =
+    rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type());
   for (size_t update_counter = 0; update_counter <= 10 * cm_update_rate; ++update_counter)
   {
     EXPECT_EQ(

--- a/controller_manager/test/test_controller_manager_hardware_error_handling.cpp
+++ b/controller_manager/test/test_controller_manager_hardware_error_handling.cpp
@@ -138,11 +138,7 @@ public:
     cm_->configure_controller(TEST_BROADCASTER_ALL_NAME);
     cm_->configure_controller(TEST_BROADCASTER_SENSOR_NAME);
 
-    EXPECT_EQ(
-      controller_interface::return_type::OK,
-      cm_->update(
-        rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-        PERIOD));
+    EXPECT_EQ(controller_interface::return_type::OK, cm_->update(time_, PERIOD));
     EXPECT_EQ(0u, test_controller_actuator->internal_counter) << "Controller is not started";
     EXPECT_EQ(0u, test_controller_system->internal_counter) << "Controller is not started";
     EXPECT_EQ(0u, test_broadcaster_all->internal_counter) << "Controller is not started";
@@ -230,11 +226,7 @@ TEST_P(TestControllerManagerWithTestableCM, stop_controllers_on_hardware_read_er
     hardware_interface::lifecycle_state_names::ACTIVE);
 
   {
-    EXPECT_EQ(
-      controller_interface::return_type::OK,
-      cm_->update(
-        rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-        PERIOD));
+    EXPECT_EQ(controller_interface::return_type::OK, cm_->update(time_, PERIOD));
     EXPECT_GE(test_controller_actuator->internal_counter, 1u)
       << "Controller is started at the end of update";
     EXPECT_GE(test_controller_system->internal_counter, 1u)
@@ -257,11 +249,7 @@ TEST_P(TestControllerManagerWithTestableCM, stop_controllers_on_hardware_read_er
   // Execute first time without any errors
   {
     auto new_counter = test_controller_actuator->internal_counter + 1;
-    EXPECT_EQ(
-      controller_interface::return_type::OK,
-      cm_->update(
-        rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-        PERIOD));
+    EXPECT_EQ(controller_interface::return_type::OK, cm_->update(time_, PERIOD));
     EXPECT_EQ(test_controller_actuator->internal_counter, new_counter) << "Execute without errors";
     EXPECT_EQ(test_controller_system->internal_counter, new_counter) << "Execute without errors";
     EXPECT_EQ(test_broadcaster_all->internal_counter, new_counter) << "Execute without errors";
@@ -273,11 +261,7 @@ TEST_P(TestControllerManagerWithTestableCM, stop_controllers_on_hardware_read_er
   test_controller_actuator->set_first_command_interface_value_to = READ_FAIL_VALUE;
   {
     auto new_counter = test_controller_actuator->internal_counter + 1;
-    EXPECT_EQ(
-      controller_interface::return_type::OK,
-      cm_->update(
-        rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-        PERIOD));
+    EXPECT_EQ(controller_interface::return_type::OK, cm_->update(time_, PERIOD));
     EXPECT_EQ(test_controller_actuator->internal_counter, new_counter)
       << "Execute without errors to write value";
     EXPECT_EQ(test_controller_system->internal_counter, new_counter)
@@ -294,8 +278,7 @@ TEST_P(TestControllerManagerWithTestableCM, stop_controllers_on_hardware_read_er
 
     // here happens error in hardware and
     // "actuator controller" and "broadcaster all" are deactivated
-    EXPECT_NO_THROW(cm_->read(
-      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()), PERIOD));
+    EXPECT_NO_THROW(cm_->read(time_, PERIOD));
     EXPECT_EQ(
       lifecycle_msgs::msg::State::PRIMARY_STATE_INACTIVE,
       test_controller_actuator->get_state().id());
@@ -306,11 +289,7 @@ TEST_P(TestControllerManagerWithTestableCM, stop_controllers_on_hardware_read_er
     EXPECT_EQ(
       lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE, test_broadcaster_sensor->get_state().id());
 
-    EXPECT_EQ(
-      controller_interface::return_type::OK,
-      cm_->update(
-        rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-        PERIOD));
+    EXPECT_EQ(controller_interface::return_type::OK, cm_->update(time_, PERIOD));
     EXPECT_EQ(test_controller_actuator->internal_counter, previous_counter)
       << "Execute has read error and it is not updated";
     EXPECT_EQ(test_controller_system->internal_counter, new_counter)
@@ -354,11 +333,7 @@ TEST_P(TestControllerManagerWithTestableCM, stop_controllers_on_hardware_read_er
     auto previous_counter_lower = test_controller_actuator->internal_counter + 1;
     auto previous_counter_higher = test_controller_system->internal_counter + 1;
 
-    EXPECT_EQ(
-      controller_interface::return_type::OK,
-      cm_->update(
-        rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-        PERIOD));
+    EXPECT_EQ(controller_interface::return_type::OK, cm_->update(time_, PERIOD));
 
     EXPECT_EQ(test_controller_actuator->internal_counter, previous_counter_lower)
       << "Execute without errors to write value";
@@ -375,8 +350,7 @@ TEST_P(TestControllerManagerWithTestableCM, stop_controllers_on_hardware_read_er
     auto previous_counter_higher = test_controller_system->internal_counter;
     auto new_counter = test_broadcaster_sensor->internal_counter + 1;
 
-    EXPECT_NO_THROW(cm_->read(
-      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()), PERIOD));
+    EXPECT_NO_THROW(cm_->read(time_, PERIOD));
     EXPECT_EQ(
       lifecycle_msgs::msg::State::PRIMARY_STATE_INACTIVE,
       test_controller_actuator->get_state().id());
@@ -387,11 +361,7 @@ TEST_P(TestControllerManagerWithTestableCM, stop_controllers_on_hardware_read_er
     EXPECT_EQ(
       lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE, test_broadcaster_sensor->get_state().id());
 
-    EXPECT_EQ(
-      controller_interface::return_type::OK,
-      cm_->update(
-        rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-        PERIOD));
+    EXPECT_EQ(controller_interface::return_type::OK, cm_->update(time_, PERIOD));
     EXPECT_EQ(test_controller_actuator->internal_counter, previous_counter_lower)
       << "Execute has read error and it is not updated";
     EXPECT_EQ(test_controller_system->internal_counter, previous_counter_higher)
@@ -448,11 +418,7 @@ TEST_P(TestControllerManagerWithTestableCM, stop_controllers_on_hardware_write_e
     hardware_interface::lifecycle_state_names::ACTIVE);
 
   {
-    EXPECT_EQ(
-      controller_interface::return_type::OK,
-      cm_->update(
-        rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-        PERIOD));
+    EXPECT_EQ(controller_interface::return_type::OK, cm_->update(time_, PERIOD));
     EXPECT_GE(test_controller_actuator->internal_counter, 1u)
       << "Controller is started at the end of update";
     EXPECT_GE(test_controller_system->internal_counter, 1u)
@@ -475,11 +441,7 @@ TEST_P(TestControllerManagerWithTestableCM, stop_controllers_on_hardware_write_e
   // Execute first time without any errors
   {
     auto new_counter = test_controller_actuator->internal_counter + 1;
-    EXPECT_EQ(
-      controller_interface::return_type::OK,
-      cm_->update(
-        rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-        PERIOD));
+    EXPECT_EQ(controller_interface::return_type::OK, cm_->update(time_, PERIOD));
     EXPECT_EQ(test_controller_actuator->internal_counter, new_counter) << "Execute without errors";
     EXPECT_EQ(test_controller_system->internal_counter, new_counter) << "Execute without errors";
     EXPECT_EQ(test_broadcaster_all->internal_counter, new_counter) << "Execute without errors";
@@ -491,11 +453,7 @@ TEST_P(TestControllerManagerWithTestableCM, stop_controllers_on_hardware_write_e
   test_controller_actuator->set_first_command_interface_value_to = WRITE_FAIL_VALUE;
   {
     auto new_counter = test_controller_actuator->internal_counter + 1;
-    EXPECT_EQ(
-      controller_interface::return_type::OK,
-      cm_->update(
-        rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-        PERIOD));
+    EXPECT_EQ(controller_interface::return_type::OK, cm_->update(time_, PERIOD));
     EXPECT_EQ(test_controller_actuator->internal_counter, new_counter)
       << "Execute without errors to write value";
     EXPECT_EQ(test_controller_system->internal_counter, new_counter)
@@ -512,8 +470,7 @@ TEST_P(TestControllerManagerWithTestableCM, stop_controllers_on_hardware_write_e
 
     // here happens error in hardware and
     // "actuator controller" and "broadcaster all" are deactivated
-    EXPECT_NO_THROW(cm_->write(
-      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()), PERIOD));
+    EXPECT_NO_THROW(cm_->write(time_, PERIOD));
     EXPECT_EQ(
       lifecycle_msgs::msg::State::PRIMARY_STATE_INACTIVE,
       test_controller_actuator->get_state().id());
@@ -524,11 +481,7 @@ TEST_P(TestControllerManagerWithTestableCM, stop_controllers_on_hardware_write_e
     EXPECT_EQ(
       lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE, test_broadcaster_sensor->get_state().id());
 
-    EXPECT_EQ(
-      controller_interface::return_type::OK,
-      cm_->update(
-        rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-        PERIOD));
+    EXPECT_EQ(controller_interface::return_type::OK, cm_->update(time_, PERIOD));
     EXPECT_EQ(test_controller_actuator->internal_counter, previous_counter)
       << "Execute without errors to write value";
     EXPECT_EQ(test_controller_system->internal_counter, new_counter)
@@ -572,11 +525,7 @@ TEST_P(TestControllerManagerWithTestableCM, stop_controllers_on_hardware_write_e
     auto previous_counter_lower = test_controller_actuator->internal_counter + 1;
     auto previous_counter_higher = test_controller_system->internal_counter + 1;
 
-    EXPECT_EQ(
-      controller_interface::return_type::OK,
-      cm_->update(
-        rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-        PERIOD));
+    EXPECT_EQ(controller_interface::return_type::OK, cm_->update(time_, PERIOD));
 
     EXPECT_EQ(test_controller_actuator->internal_counter, previous_counter_lower)
       << "Execute without errors to write value";
@@ -595,8 +544,7 @@ TEST_P(TestControllerManagerWithTestableCM, stop_controllers_on_hardware_write_e
 
     // here happens error in hardware and
     // "actuator controller" and "broadcaster all" are deactivated
-    EXPECT_NO_THROW(cm_->write(
-      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()), PERIOD));
+    EXPECT_NO_THROW(cm_->write(time_, PERIOD));
     EXPECT_EQ(
       lifecycle_msgs::msg::State::PRIMARY_STATE_INACTIVE,
       test_controller_actuator->get_state().id());
@@ -607,11 +555,7 @@ TEST_P(TestControllerManagerWithTestableCM, stop_controllers_on_hardware_write_e
     EXPECT_EQ(
       lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE, test_broadcaster_sensor->get_state().id());
 
-    EXPECT_EQ(
-      controller_interface::return_type::OK,
-      cm_->update(
-        rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-        PERIOD));
+    EXPECT_EQ(controller_interface::return_type::OK, cm_->update(time_, PERIOD));
     EXPECT_EQ(test_controller_actuator->internal_counter, previous_counter_lower)
       << "Execute has write error and it is not updated";
     EXPECT_EQ(test_controller_system->internal_counter, previous_counter_higher)

--- a/controller_manager/test/test_controller_manager_hardware_error_handling.cpp
+++ b/controller_manager/test/test_controller_manager_hardware_error_handling.cpp
@@ -138,7 +138,11 @@ public:
     cm_->configure_controller(TEST_BROADCASTER_ALL_NAME);
     cm_->configure_controller(TEST_BROADCASTER_SENSOR_NAME);
 
-    EXPECT_EQ(controller_interface::return_type::OK, cm_->update(TIME, PERIOD));
+    EXPECT_EQ(
+      controller_interface::return_type::OK,
+      cm_->update(
+        rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+        PERIOD));
     EXPECT_EQ(0u, test_controller_actuator->internal_counter) << "Controller is not started";
     EXPECT_EQ(0u, test_controller_system->internal_counter) << "Controller is not started";
     EXPECT_EQ(0u, test_broadcaster_all->internal_counter) << "Controller is not started";
@@ -226,7 +230,11 @@ TEST_P(TestControllerManagerWithTestableCM, stop_controllers_on_hardware_read_er
     hardware_interface::lifecycle_state_names::ACTIVE);
 
   {
-    EXPECT_EQ(controller_interface::return_type::OK, cm_->update(TIME, PERIOD));
+    EXPECT_EQ(
+      controller_interface::return_type::OK,
+      cm_->update(
+        rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+        PERIOD));
     EXPECT_GE(test_controller_actuator->internal_counter, 1u)
       << "Controller is started at the end of update";
     EXPECT_GE(test_controller_system->internal_counter, 1u)
@@ -249,7 +257,11 @@ TEST_P(TestControllerManagerWithTestableCM, stop_controllers_on_hardware_read_er
   // Execute first time without any errors
   {
     auto new_counter = test_controller_actuator->internal_counter + 1;
-    EXPECT_EQ(controller_interface::return_type::OK, cm_->update(TIME, PERIOD));
+    EXPECT_EQ(
+      controller_interface::return_type::OK,
+      cm_->update(
+        rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+        PERIOD));
     EXPECT_EQ(test_controller_actuator->internal_counter, new_counter) << "Execute without errors";
     EXPECT_EQ(test_controller_system->internal_counter, new_counter) << "Execute without errors";
     EXPECT_EQ(test_broadcaster_all->internal_counter, new_counter) << "Execute without errors";
@@ -261,7 +273,11 @@ TEST_P(TestControllerManagerWithTestableCM, stop_controllers_on_hardware_read_er
   test_controller_actuator->set_first_command_interface_value_to = READ_FAIL_VALUE;
   {
     auto new_counter = test_controller_actuator->internal_counter + 1;
-    EXPECT_EQ(controller_interface::return_type::OK, cm_->update(TIME, PERIOD));
+    EXPECT_EQ(
+      controller_interface::return_type::OK,
+      cm_->update(
+        rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+        PERIOD));
     EXPECT_EQ(test_controller_actuator->internal_counter, new_counter)
       << "Execute without errors to write value";
     EXPECT_EQ(test_controller_system->internal_counter, new_counter)
@@ -278,7 +294,8 @@ TEST_P(TestControllerManagerWithTestableCM, stop_controllers_on_hardware_read_er
 
     // here happens error in hardware and
     // "actuator controller" and "broadcaster all" are deactivated
-    EXPECT_NO_THROW(cm_->read(TIME, PERIOD));
+    EXPECT_NO_THROW(cm_->read(
+      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()), PERIOD));
     EXPECT_EQ(
       lifecycle_msgs::msg::State::PRIMARY_STATE_INACTIVE,
       test_controller_actuator->get_state().id());
@@ -289,7 +306,11 @@ TEST_P(TestControllerManagerWithTestableCM, stop_controllers_on_hardware_read_er
     EXPECT_EQ(
       lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE, test_broadcaster_sensor->get_state().id());
 
-    EXPECT_EQ(controller_interface::return_type::OK, cm_->update(TIME, PERIOD));
+    EXPECT_EQ(
+      controller_interface::return_type::OK,
+      cm_->update(
+        rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+        PERIOD));
     EXPECT_EQ(test_controller_actuator->internal_counter, previous_counter)
       << "Execute has read error and it is not updated";
     EXPECT_EQ(test_controller_system->internal_counter, new_counter)
@@ -333,7 +354,11 @@ TEST_P(TestControllerManagerWithTestableCM, stop_controllers_on_hardware_read_er
     auto previous_counter_lower = test_controller_actuator->internal_counter + 1;
     auto previous_counter_higher = test_controller_system->internal_counter + 1;
 
-    EXPECT_EQ(controller_interface::return_type::OK, cm_->update(TIME, PERIOD));
+    EXPECT_EQ(
+      controller_interface::return_type::OK,
+      cm_->update(
+        rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+        PERIOD));
 
     EXPECT_EQ(test_controller_actuator->internal_counter, previous_counter_lower)
       << "Execute without errors to write value";
@@ -350,7 +375,8 @@ TEST_P(TestControllerManagerWithTestableCM, stop_controllers_on_hardware_read_er
     auto previous_counter_higher = test_controller_system->internal_counter;
     auto new_counter = test_broadcaster_sensor->internal_counter + 1;
 
-    EXPECT_NO_THROW(cm_->read(TIME, PERIOD));
+    EXPECT_NO_THROW(cm_->read(
+      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()), PERIOD));
     EXPECT_EQ(
       lifecycle_msgs::msg::State::PRIMARY_STATE_INACTIVE,
       test_controller_actuator->get_state().id());
@@ -361,7 +387,11 @@ TEST_P(TestControllerManagerWithTestableCM, stop_controllers_on_hardware_read_er
     EXPECT_EQ(
       lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE, test_broadcaster_sensor->get_state().id());
 
-    EXPECT_EQ(controller_interface::return_type::OK, cm_->update(TIME, PERIOD));
+    EXPECT_EQ(
+      controller_interface::return_type::OK,
+      cm_->update(
+        rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+        PERIOD));
     EXPECT_EQ(test_controller_actuator->internal_counter, previous_counter_lower)
       << "Execute has read error and it is not updated";
     EXPECT_EQ(test_controller_system->internal_counter, previous_counter_higher)
@@ -418,7 +448,11 @@ TEST_P(TestControllerManagerWithTestableCM, stop_controllers_on_hardware_write_e
     hardware_interface::lifecycle_state_names::ACTIVE);
 
   {
-    EXPECT_EQ(controller_interface::return_type::OK, cm_->update(TIME, PERIOD));
+    EXPECT_EQ(
+      controller_interface::return_type::OK,
+      cm_->update(
+        rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+        PERIOD));
     EXPECT_GE(test_controller_actuator->internal_counter, 1u)
       << "Controller is started at the end of update";
     EXPECT_GE(test_controller_system->internal_counter, 1u)
@@ -441,7 +475,11 @@ TEST_P(TestControllerManagerWithTestableCM, stop_controllers_on_hardware_write_e
   // Execute first time without any errors
   {
     auto new_counter = test_controller_actuator->internal_counter + 1;
-    EXPECT_EQ(controller_interface::return_type::OK, cm_->update(TIME, PERIOD));
+    EXPECT_EQ(
+      controller_interface::return_type::OK,
+      cm_->update(
+        rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+        PERIOD));
     EXPECT_EQ(test_controller_actuator->internal_counter, new_counter) << "Execute without errors";
     EXPECT_EQ(test_controller_system->internal_counter, new_counter) << "Execute without errors";
     EXPECT_EQ(test_broadcaster_all->internal_counter, new_counter) << "Execute without errors";
@@ -453,7 +491,11 @@ TEST_P(TestControllerManagerWithTestableCM, stop_controllers_on_hardware_write_e
   test_controller_actuator->set_first_command_interface_value_to = WRITE_FAIL_VALUE;
   {
     auto new_counter = test_controller_actuator->internal_counter + 1;
-    EXPECT_EQ(controller_interface::return_type::OK, cm_->update(TIME, PERIOD));
+    EXPECT_EQ(
+      controller_interface::return_type::OK,
+      cm_->update(
+        rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+        PERIOD));
     EXPECT_EQ(test_controller_actuator->internal_counter, new_counter)
       << "Execute without errors to write value";
     EXPECT_EQ(test_controller_system->internal_counter, new_counter)
@@ -470,7 +512,8 @@ TEST_P(TestControllerManagerWithTestableCM, stop_controllers_on_hardware_write_e
 
     // here happens error in hardware and
     // "actuator controller" and "broadcaster all" are deactivated
-    EXPECT_NO_THROW(cm_->write(TIME, PERIOD));
+    EXPECT_NO_THROW(cm_->write(
+      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()), PERIOD));
     EXPECT_EQ(
       lifecycle_msgs::msg::State::PRIMARY_STATE_INACTIVE,
       test_controller_actuator->get_state().id());
@@ -481,7 +524,11 @@ TEST_P(TestControllerManagerWithTestableCM, stop_controllers_on_hardware_write_e
     EXPECT_EQ(
       lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE, test_broadcaster_sensor->get_state().id());
 
-    EXPECT_EQ(controller_interface::return_type::OK, cm_->update(TIME, PERIOD));
+    EXPECT_EQ(
+      controller_interface::return_type::OK,
+      cm_->update(
+        rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+        PERIOD));
     EXPECT_EQ(test_controller_actuator->internal_counter, previous_counter)
       << "Execute without errors to write value";
     EXPECT_EQ(test_controller_system->internal_counter, new_counter)
@@ -525,7 +572,11 @@ TEST_P(TestControllerManagerWithTestableCM, stop_controllers_on_hardware_write_e
     auto previous_counter_lower = test_controller_actuator->internal_counter + 1;
     auto previous_counter_higher = test_controller_system->internal_counter + 1;
 
-    EXPECT_EQ(controller_interface::return_type::OK, cm_->update(TIME, PERIOD));
+    EXPECT_EQ(
+      controller_interface::return_type::OK,
+      cm_->update(
+        rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+        PERIOD));
 
     EXPECT_EQ(test_controller_actuator->internal_counter, previous_counter_lower)
       << "Execute without errors to write value";
@@ -544,7 +595,8 @@ TEST_P(TestControllerManagerWithTestableCM, stop_controllers_on_hardware_write_e
 
     // here happens error in hardware and
     // "actuator controller" and "broadcaster all" are deactivated
-    EXPECT_NO_THROW(cm_->write(TIME, PERIOD));
+    EXPECT_NO_THROW(cm_->write(
+      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()), PERIOD));
     EXPECT_EQ(
       lifecycle_msgs::msg::State::PRIMARY_STATE_INACTIVE,
       test_controller_actuator->get_state().id());
@@ -555,7 +607,11 @@ TEST_P(TestControllerManagerWithTestableCM, stop_controllers_on_hardware_write_e
     EXPECT_EQ(
       lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE, test_broadcaster_sensor->get_state().id());
 
-    EXPECT_EQ(controller_interface::return_type::OK, cm_->update(TIME, PERIOD));
+    EXPECT_EQ(
+      controller_interface::return_type::OK,
+      cm_->update(
+        rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+        PERIOD));
     EXPECT_EQ(test_controller_actuator->internal_counter, previous_counter_lower)
       << "Execute has write error and it is not updated";
     EXPECT_EQ(test_controller_system->internal_counter, previous_counter_higher)

--- a/controller_manager/test/test_controllers_chaining_with_controller_manager.cpp
+++ b/controller_manager/test/test_controllers_chaining_with_controller_manager.cpp
@@ -378,7 +378,9 @@ public:
     position_tracking_controller->external_commands_for_testing_[0] = reference[0];
     position_tracking_controller->external_commands_for_testing_[1] = reference[1];
 
-    cm_->update(rclcpp::Time(0), rclcpp::Duration::from_seconds(0.01));
+    cm_->update(
+      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+      rclcpp::Duration::from_seconds(0.01));
     cm_->resource_manager_->read(rclcpp::Time(0), rclcpp::Duration::from_seconds(0.01));
 
     // check if all controllers are updated

--- a/controller_manager/test/test_controllers_chaining_with_controller_manager.cpp
+++ b/controller_manager/test/test_controllers_chaining_with_controller_manager.cpp
@@ -378,9 +378,7 @@ public:
     position_tracking_controller->external_commands_for_testing_[0] = reference[0];
     position_tracking_controller->external_commands_for_testing_[1] = reference[1];
 
-    cm_->update(
-      rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-      rclcpp::Duration::from_seconds(0.01));
+    cm_->update(time_, rclcpp::Duration::from_seconds(0.01));
     cm_->resource_manager_->read(rclcpp::Time(0), rclcpp::Duration::from_seconds(0.01));
 
     // check if all controllers are updated

--- a/controller_manager/test/test_spawner_unspawner.cpp
+++ b/controller_manager/test/test_spawner_unspawner.cpp
@@ -39,15 +39,9 @@ class TestLoadController : public ControllerManagerFixture<controller_manager::C
       std::chrono::milliseconds(10),
       [&]()
       {
-        cm_->read(
-          rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-          PERIOD);
-        cm_->update(
-          rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-          PERIOD);
-        cm_->write(
-          rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
-          PERIOD);
+        cm_->read(time_, PERIOD);
+        cm_->update(time_, PERIOD);
+        cm_->write(time_, PERIOD);
       });
 
     update_executor_ =

--- a/controller_manager/test/test_spawner_unspawner.cpp
+++ b/controller_manager/test/test_spawner_unspawner.cpp
@@ -39,9 +39,15 @@ class TestLoadController : public ControllerManagerFixture<controller_manager::C
       std::chrono::milliseconds(10),
       [&]()
       {
-        cm_->read(TIME, PERIOD);
-        cm_->update(TIME, PERIOD);
-        cm_->write(TIME, PERIOD);
+        cm_->read(
+          rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+          PERIOD);
+        cm_->update(
+          rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+          PERIOD);
+        cm_->write(
+          rclcpp::Time(0, 0, cm_->get_node_clock_interface()->get_clock()->get_clock_type()),
+          PERIOD);
       });
 
     update_executor_ =

